githubkit/core.py

```diff
@@ -391,6 +403,33 @@ def _extract_retry_after(self, response: Response) -> timedelta:
             # wait for at least one minute before retrying
             return timedelta(seconds=60)
 
+    def _handle_retry(self, exc: GitHubException, retry_count: int) -> RetryOption:
+        # retry once for rate limit exceeded
+        # https://github.com/octokit/octokit.js/blob/450c662e4f29b63a6dbe7592ba5ef53d5fa01d88/src/octokit.ts#L34-L65
+        if retry_count < 1 and isinstance(exc, RateLimitExceeded):
+            logger.warning(
+                f"{exc.__class__.__name__} for request "
+                f"{exc.request.method} {exc.request.url}",
+                exc_info=exc,
+            )
+            return RetryOption(True, exc.retry_after)
+
+        # retry three times for server error
+        # https://github.com/octokit/plugin-retry.js/blob/3b9897a58d8b2c37cfba6a8df78e4619d346098a/src/error-request.ts#L11-L14
+        if (
+            retry_count < 3
+            and isinstance(exc, RequestFailed)
+            and exc.response.status_code >= 500
+        ):
+            logger.warning(
+                f"Server error encountered for request "
+                f"{exc.request.method} {exc.request.url}",
+                exc_info=exc,
+            )
+            return RetryOption(True, timedelta(seconds=(retry_count + 1) ** 2))
+
+        return RetryOption(False)
```

Comment:
It looks good!

Maybe it's a good idea to move/extract this method from the class.
It is not really connected to the rest of the class (`self` is not used for example).
Afterwards it can be made the default value for `auto_retry`.
If you make it into a callable class, you could even turn the hard coded values into (default) initialization parameters.

This could also simplify the type of `auto_retry` (not a mixture of types, but `Callable[...]|None`) and the logic in line 465-470 (check if None instead of false).

A final suggestion would be to rename `RetryHandler` to something else, since it does not actually handle the retry attempt. I cannot think of a better name though😅

Reply:
> A final suggestion would be to rename RetryHandler to something else

Maybe `RetryDecisionFunc`? (GPT suggests this)

Reply:
> This could also simplify the type of `auto_retry` (not a mixture of types, but `Callable[...]|None`) and the logic in line 465-470 (check if None instead of false).

If remove the `bool` type, user needs to use `None` to represent disabling this feature , and the arg default value will come to be a callable. This does not looks good. I think this mixture of type can be handled when building config.

Reply:
@schelv i have update the impl, could you please make a review?